enhancements/implicit-catalog-versioning.md

```diff
@@ -0,0 +1,190 @@
+
+In the future, this proposal also helps to solve the ordering problem when building and modifying index images. Today when adding to Index images via the `opm index` command, you need to specify your images in the order that they replace previous images because of the possibility that the explicit data will make an invalid intermediate graph. With the use of semantic version, it becomes trivial to insert elements into the middle of the graph. It also allows a future in which specific versions can be deleted from graphs without a need to edit the yaml.
+
+### Goals
+
+* Define Catalog update graphs for individual operators based implicitly on semantic version rather than asking users to explicitly mention what version the previous version replaces
+* Update OLM to handle the new format of catalogs
+* Handle conversion of existing operator subscriptions to use the new format
+
+### Non-Goals
+
+* Modify any of the features of OLM or the Operator Registry to use new information from the semantic version. This proposal simply suggests using the semantic version to define the graph
+* Resolve the migration of existing subscriptions to use the new Catalog Index images
+* Make any explicit changes to existing OLM API versions (CSV will remain `v1alpha1`)
+
+## Proposal
+
+Basic approach:
+
+- Make the `spec.version` field required going forward when building registry images with `opm registry` commands
+- Modify the operator registry to build dependency graph based simply on `spec.version` field rather than using a combination of channel heads and `spec.replaces` for each graph
+- Modify OLM to use the `spec.version` field everywhere it attempts to parse the graph for update information
+- Update skip feature to use version strings
+
+### Deprecate the `replaces` field:
+
+Currently, the update graph defined when creating an operator catalog is explicit. Essentially, the ClusterServiceVersion (CSV) file has an explicit `replaces` field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+    replaces: etcdoperator.v0.9.0
+```
+
+The `replaces` field is the source for the update graph for each install channel in a given operator. When the `operator-registry` builds the update graph, it actually walks through each CSV in a given Operator and, for each channel, constructs a chain of CSVs that define the update path. Instead, we can just define our spec without the replaces field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+```
+
+We would use the fact that 0.9.2 > 0.9.0 to implicitly assume that 0.9.0 is replaced by 0.9.2 through the rules of semantic version.
```

Comment:
This is good, but I think we need more information.

What are the "upgrade rules" that we will apply? What happens when I insert a patch release for an older package? Can I mark a release as a required upgrade step? 

Do we still have channels at all? Are they based on versions (i.e. minor version channels)? 

Is there a proposal to allow subscriptions to specify allowed upgrade ranges like some package managers allow? (i.e. `~1.2.3`) Or does the operator author retain control of upgrades?

Reply:
I can update this to be more detailed. Some specific responses below:

Essentially, this just makes it possible to easily insert minor and patch updates into the middle of the upgrade graph based on semantic versioning rules (`1.2.3` replaces `1.2.2` replaces `1.1.0` replaces `0.9.0`)

OLM still steps through updates unless you explicitly specify a `skipRange` in your CSV which is still essentially a list of versions that can update directly to that CSV.

Channels still exist. Each bundle explicitly subscribes to a set of channels. Minor openshift versions specifically are (in the current proposal) going to create separate registry images and no longer solve the minor versioning problem with channels. So they could still be used to just define arbitrary update paths (`alpha`, `$experimentalFeature`, `stable` etc.)

There is not right now, but this does make that seem much simpler to implement.

Reply:
> OLM still steps through updates...

We should think about whether this is what users expect when using semver. If the answer is no, and users are unable to select versions directly, then I don't think it makes sense to support inserting patch versions.

Reply:
> We should think about whether this is what users expect when using semver

This is my only remaining reservation with the enhancement. It is a big UX win to be able to insert into older parts of the graph easily, but do people expect to have to upgrade through every patch version?

Even that seems fine as long as we have a migration strategy (i.e. we can easily provide alternate commands or options for adding to the index that do other things)